weasyl/two_factor_auth.py

```diff
@@ -0,0 +1,349 @@
+"""
+Module for handling 2FA-related functions.
+"""
+from __future__ import absolute_import, unicode_literals
+
+import re
+import urllib
+
+import bcrypt
+import pyotp
+from qrcodegen import QrCode
+
+from libweasyl import security
+from weasyl import define as d
+from weasyl import login
+
+# Number of recovery codes to provide the user
+_TFA_RECOVERY_CODES = 10
+# Work factor for bcrypt to be used in hashing each recovery code
+BCRYPT_WORK_FACTOR = 10
```

Comment:
As a note, I am not sure how much time we want to spend during the setup to generate the bcrypt hashes; I will be happy to change it to a higher value if anyone thinks it should be done (bearing in mind that the main TOTP secret is still unencrypted in ``login``). The overall initial cost would be however long it takes to generate the hash at N-work factor, times the recovery codes provided, which would make a bit of a sizable page load delay during the ``tfa_init_verify_post_()`` controller step, at least depending on how much we want to secure these.

As a comparison, the default work factor bcrypt uses is 12, and Weasyl passwords are 14 (as of this commit, and controlled by a variable in the ``macro.py`` module).

I'll leave it up to the reviewers and other Weasyl staff to determine how long of a page-load is tolerable (though in this case it may be justified, especially if we make note of it in the page itself).